src/NuGet.Core/NuGet.Packaging/PackageExtraction/StreamExtensions.cs

```diff
@@ -15,6 +16,34 @@ public static class StreamExtensions
         */
         private const long MAX_MMAP_SIZE = 10 * 1024 * 1024;
 
+        // Mmap can improve file writing performance, but it can make it slower too.
+        // It all depends on a particular hardware configuration, operating system or anti-virus software.
+        // From our benchmarks we concluded that mmap is a good choice for Windows,
+        // but it is not so for other systems.
+        //
+        // 1 - always use memory-mapped files
+        // 0 - never use memory-mapped files
+        // default - use memory-mapped files on Windows, but not on other systems
+        private const string MMAP_VARIABLE_NAME = "NUGET_PACKAGE_EXTRACTION_USE_MMAP";
+        private static bool IsMMapEnabled { get; }
+
+        static StreamExtensions()
```

Comment:
@zivkan Might be a good idea to make this testable.

Reply:
@nkolev92  I'm trying to think what/how to test.

Regardless of the env var value, the code is supposed to write files to disk. So checking that the file is written to disk with the correct contents will not validate that the correct extraction method is used.

trying to measure the duration of the extraction to guess whether mmap or FileStream was used on each platform sounds extremely flakey and prone to machine state (on CI we can be confident machine state will be consistent, but I think everyone should be able to run tests locally as well).

We could add EventSource events, and then have the test use a EventListener, but I really don't want to do that unless we think the ETW events will be useful for perf investigations, not just testing. We could write a test that uses EventListener without adding any EventSource to our production code, and with debugging try to find if the .NET runtime has any of its own ETW events that we could use to validate. But that feels like our tests would use undocumented APIs/events that could change at any time.

So in the principal of "everything should be tested", ok. But I'm not sure what we'd actually be testing.

Or do you just mean test that the default is true on WIndows and false on everything else, but not validate that the boolean is used correctly?

To be clear to @marcin-krystianc we won't block the PR on tests, it's something we'd do ourselves after this is merged.

Reply:
I'd just test as much as reasonable. 

So testing that it's true on windows and false on others would be good. 
Testing that the env var is respected is also a good one.

Reply:
> To be clear to @marcin-krystianc we won't block the PR on tests, it's something we'd do ourselves after this is merged.

No worries, I am happy to add some tests.  I can easily make `IsMMapEnabled` internal so it can be checked whether it is `true` on Windows and false on other systems. 
Testing whether the env variable is respected is more tricky though since `IsMMapEnabled` is static. By making it non-nonstatic we would make it more testable, but on the other hand, the env variable would be checked each time a file is extracted. Not sure whether it is worth the extra cost at runtime.
So, should I just stick to the "true on windows and false on other systems" testing?

Reply:
In https://github.com/NuGet/NuGet.Client/pull/4377, I had a static class add an internal (non-static) class that is testable, so a similar approach could be used. By using a static instance of the non-static class in production code, it ensures the environment varible is only read & parsed only once, but the internal class is testable with our abstracted environment variable interface.

Maybe you have a better idea than me, but I just had the idea to make MmapCopy and FileStreamCopy virtual methods, so the test overwrite the implement the implementation to set a bool, which the test validates. I don't know how I feel about that idea, but it does open the possibility to a whole bunch of tests: Validate that the correct method is used, depending on the value of IsMmapEnabled. Validate that CopyTo(FileStream) writes the correct contents into the destination file. Validate that CopyTo(MmapStream) writes the correct contents into the destination file (we have a very small number of bug reports that some extract packages have nupec files with corrupted contents, and I think it might have started after mmap was added).

Reply:
@zivkan I've followed your advice and extracted the code we want to test into a new class. Please have another look, let me know if there is anything to add.